stubs/setuptools/setuptools/_distutils/extension.pyi

```diff
@@ -18,7 +20,7 @@ class Extension:
     def __init__(
         self,
         name: str,
-        sources: list[str],
+        sources: list[str] | list[StrPath],
```

Comment:
Yes, if `Extension.sources` retrieves a list anyway, requiring a list in the input seems like it's just an unnecessary restriction. The long-term solution with generics is interesting and makes (somewhat) sense to me, though – would you like to see me try a follow-up PR to pypa/distutils with that approach sometime soon?

Reply:
I don't think any change to pypa/distutils is necessary to make this class generic in stubs.

Whilst it's currently not pretty, I'd be fine doing it as a follow-up and keeping this PR as-is to restore functionality to projects that were relying on the previous annotations.

I'm curious as to what other maintainers think though. Is the current state of this PR acceptable as a stop-gap with a follow-up? Or should we make it generic now? Other thoughts? (CCing @hauntsaninja from conversation in #18107 )

Reply:
I'd personally prefer to avoid the generic. Generics are confusing for users who use the class in a type annotation, because they need to figure out what the generic parameter refers to. That's easy for something where the generic nature is really inherent to the class (like with `list` or other collections), but for this class, the generic parameter probably doesn't make a huge difference to most users. It may be better to go with a more general type like Sequence to improve usability.

Reply:
Also I wrote this before I saw @Avasam's comment above. I think the current state of this PR is OK; the one thing that gives me pause is the `list[PathLike[str]] | list[str] ...` annotation. Type checkers aren't terribly likely to infer `list[PathLike[str]]` as the type for a list that the user provides, so that type isn't very usable in practice. We could consider using `Sequence` instead, but then we'd get false negatives if a user passes a tuple or other sequence. Ideally we'd solve this in the runtime of setuptools so it accepts any Sequence and not just list, but in the meantime what the PR currently does seems like a good solution.